src/contract.rs

```diff
@@ -73,9 +76,120 @@ pub fn execute(
             validator_oper_addr,
         } => remove_validator(deps, info, validator_oper_addr),
         ExecuteMsg::WithdrawFunds { amount } => withdraw_funds(deps, info, amount),
+
+        ExecuteMsg::DeleteMetricsForTimestamp {
+            timestamp,
+            validator_idx,
+            validator_ct,
+        } => delete_metrics_for_timestamp(
+            deps,
+            info,
+            timestamp,
+            validator_idx as usize,
+            validator_ct as usize,
+        ),
+
+        ExecuteMsg::DeleteMetricsForValidator {
+            validator_opr_addr,
+            timestamp_idx,
+            timestamp_ct,
+        } => delete_metrics_for_validator(
+            deps,
+            info,
+            validator_opr_addr,
+            timestamp_idx as usize,
+            timestamp_ct as usize,
+        ),
+
+        ExecuteMsg::RemoveTimestamp { timestamp_idx } => delete_timestamp(deps, info, timestamp_idx)
     }
 }
 
+//note: this is also inefficient, potentially O(T), due to array deletion. not sure if
+// cosmwasm library optimizes this.
+fn delete_timestamp(
+    deps: DepsMut,
+    info: MessageInfo,
+    timestamp: u64) -> Result<Response, ContractError> {
+
+    // can only be called by manager
+    if !sender_is_manager(&deps, &info) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut state = STATE.load(deps.storage)?;
+    let existing_timestamps_length = state.cron_timestamps.len();
+
+    let new_timestamps = state.cron_timestamps.into_iter()
+        .filter(|state_tstamp| {state_tstamp.ne(&timestamp)})
+        .collect::<Vec<u64>>();
+
+    if new_timestamps.len() == existing_timestamps_length {
+        return Err(ContractError::InvalidTimestamp{});
+    }
+
+    state.cron_timestamps = new_timestamps;
+    STATE.save(deps.storage, &state)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "delete_timestamp")
+        .add_attribute("timestamp_removed", timestamp.to_string()))
+}
+
+fn sender_is_manager(deps: &DepsMut, info: &MessageInfo) -> bool {
+    let config = CONFIG.load(deps.storage).unwrap();
+    info.sender == config.manager
+}
+
+fn delete_metrics_for_timestamp(
```

Comment:
Can we programmatically ensure the ordering of `remove_timestamp` and `delete_metrics_for_timestamp` somehow? No worries if there's no clean approach. Would be a nice to have.

Reply:
Yo ensure ordering, one idea I had was to have an additional data structure to store the ct of timestamps that a validator has metrics in, and vice versa.

However, this would require migration of past data and new fields in metrics / state. Would that be acceptable?

Reply:
Instead of a migration, can we not initialize the count to be the length of state.timestamps?

Reply:
Don't think so, as due to validator addition, some validators might have t timestamps' recorded metrics, and others might have (t-1), (t-x), as the validator would not have existed in list of validators for some timestamps.

Similarly, for some timestamps, validators might be v, and for others it might be (v-1), (v-x), depending on when the validators were added, as when the timestamp's cron was run some validators might not have been added.

Reply:
Right, but would you not loop through all timestamps for a validator based on state.timestamps when deleting?

Reply:
Yeah, we do. But to ensure that the ordering of  delete_metrics_for_timestamp and remove_timestamp is only allowed in correct order, we'd have to ensure that the remove_timestamp throws an error if there are metrics existing for this timestamp.

In order to do it, we'd have to ensure that there are no metrics for the timestamp we're deleting. The way to do that in O(1 - logN) time is to have a lookup for count of metrics for this timestamp, or through size of a set for this timestamp's metrics.